master/buildbot/test/util/integration.py

```diff
@@ -21,16 +21,65 @@
 
 from twisted.internet import defer
 from twisted.internet import reactor
+from twisted.internet import task
 from twisted.trial import unittest
 
 from buildbot.master import BuildMaster
 from buildbot.status.results import SUCCESS
 from buildbot.status.results import statusToString
 from buildbot.test.util import dirs
+from buildbot.schedulers.basic import BaseBasicScheduler
 from buildslave.bot import BuildSlave
 from buildslave.bot import LocalBuildSlave
 
 
+class WaitForStartConsumingSchedulers(object):
+    """
+    startService does not return a deferred. So, when
+    we call yield master.reconfigServiceWithBuildbotConfig
+    we cannot know whether services, like schedulers, have finished their
+    setup.
+    This creates an issue regarding the SingleBranch, AnyBranch schedulers.
+    These schedulers are able to trigger builds only after they have
+    subscribed to the 'new' change event. If we try to start a build with a new change
+    before they subscribed to the event (what can happen as startService does not return a deferred),
+    the build will never start, and the test will timeout.
+    This class implements a poller to ensure that these schedulers have subscribed to the event,
+    before any test tries to create a build with a new change.
+    """
+    POLL_INTERVAL_SEC = 1
```

Comment:
CusteredService: 
Thanks for the information.

Regarding the flaky decorator solution, I agree with you.
Moreover, even if we try to implement it, note that we should put it on all integration tests that trigger builds my means of <AnyBranch or SingleBranch>Scheduler + add change mechanism, which does not seem very feasible.
@sa2ajj do you agree with the proposal ?
- merge this PR,
- make the CulsteredService startService become async in another PR, that will contain a revert to one of the two commits of this one: https://github.com/yetanotherion/buildbot/commit/866e90b5bdd0d4a820c1e6b4bfa2be61cb2b9218